strongbox-security/strongbox-authentication-providers/strongbox-ldap-authentication-provider/src/test/resources/org/carlspring/strongbox/authentication/api/impl/ldap/lapt.properties

```diff
@@ -16,3 +16,4 @@ strongbox.authentication.ldap.authorities.convertToUpperCase=false
 strongbox.authentication.ldap.authorities.groupRoleAttribute=cn
 strongbox.authentication.ldap.authorities.rolePrefix=
 strongbox.authentication.ldap.authorities.searchSubtree=true
+strongbox.authentication.ldap.base64EncodedPassword=false
```

Comment:
What is this used for exactly?
Shouldn't it be able to guess this automatically?

Reply:
@carlspring we talked about this with @steve-todorov and agreed to have it like this. The thing is that it's not obvious how to guess that the string is base64 encoded.

Reply:
> The thing is that it's not obvious how to guess that the string is base64 encoded

I agree it would be tricky to auto guess indeed. The closest to auto-guessing we come to is to use fallbacks which would also allow support for both base64 and non-base64 encoded passwords. The TL;DR version:

1. Get user password from LDAP
2. Inside the `CustomLdapUserDetailsMapper`
    1. Try to compare passwords using regular Spring Password Encoder - if it doesn't succeed then `ii.`
    2. Try to compare passwords using the newly created 

Other than that, I don't see any other way we can automate it.

Reply:
@sbespalov , @madhukarbharti ,

The property approach is, unfortunately, not good enough. Let me explain why...

In large organizations, you usually have to define more than one LDAP server (different teams may be in different locations and jurisdictions of the respective organization). For example, in the corporation's European office, they might have an OpenLDAP server that is part of the farm, but, in the US office, they might be using AD (ActiveDirectory). One of the chained servers in the farm might be using base64, but the other might not and when we do it this way, this will break it for the other server.

At some point we will need to add support for more than one LDAP server and when that time comes, we'll have to re-work this, if we do it like this.

I agree with Stevie's proposal about this -- we should do our best to guess things and if `2.i` and `2.ii` don't work, well, the users of our application will (then) simply have to work it out or report an issue with the problem they are facing, so we can investigate and improve our implementation. This way we'll have done our best to support them through our implementation, so that they have less to worry about.

Reply:
@carlspring @steve-todorov @sbespalov 

Based on the current approach where we have custom UserDetailMapper we can just map a password either with what has been passed or decoded base64. If we validate the password in our custom mapper will it not be duplicate password validations best case 2 times, else 3 times? We can have password validation logic in the new custom mapper to get away from the property. It will do:
I.  Verifies password matches, then return the same password
II. If `I` fails, based on password hash it will decode and match, if matches return decoded string else return an actual hash password

Let me know if we can change this way?

Reply:
@carlspring we already able to support many LDAP servers, this is just a matter of configuration. So regarding this  `strongbox.authentication.ldap.base64EncodedPassword` property it could have different values for different LDAP instances like below:
```
//LDAP1 server configuration
strongbox.authentication.ldap1.url=ldap://ldap1:53389/dc=carlspring,dc=com
...
strongbox.authentication.ldap1.base64EncodedPassword=false

//LDAP2 server configuration
strongbox.authentication.ldap2.url=ldap://ldap2:53389/dc=carlspring,dc=com
strongbox.authentication.ldap2.base64EncodedPassword=true
```
As for @steve-todorov proposal it seems not really good to have `PasswordEncoder` instance at `CustomLdapUserDetailsMapper` and there are many reasons of this. First is that I don't understand how it suppose to work because we have no "provided" password in `LdapUserDetailsMapper` so we have nothing to compare with `PasswordEncoder`, another thing is that it seems like a bad architectural design because it brings high cohesion between LDAP mappings and authentication process and will be confusing in general.

Summarizing things I would suggest to think in the future about how we could improve this with distinction of base64 encoded passwords. As for me then manual configuration seems relevant as well and have it's own advantages.

Reply:
You are right. The correct place for such a logic would probably be in an  LdapAuthenticationManager || LdapAuthenticationProvider as a custom PasswordEncoder but from what I see we currently don't have a way to do it yet and would require a refactoring. 

Let's go with what Sergey suggests for now.

Reply:
Very well, let's proceed with this as it is.